### DISCUSSION TITLE
What about height alignment?

### DISCUSSION THREAD

**Discussion by letavocado** [katspaugh/wavesurfer.js#2810] [Ideas]
Height alignment is useful for implementing such examples.

![image](https://github.com/katspaugh/wavesurfer.js/assets/31704887/62e5eb28-9a15-45c6-bd19-9e1fc262e9bc)

![image](https://github.com/katspaugh/wavesurfer.js/assets/31704887/bbb9018e-102a-4989-9352-8c1d701e2819)

Possible solution:

from:
https://github.com/katspaugh/wavesurfer.js/blob/c8ff5a3dd4d7603dea86785b53cad3ad62707f41/src/renderer.ts#L288-L294

to:
```js
ctx.roundRect( 
   prevX * (barWidth + barGap), 
   alignEnd ? height : halfHeight - leftBarHeight, 
   barWidth, 
   leftBarHeight + (rightBarHeight || 1), 
   barRadius, 
 ) 
```

Cheers!

**Comment by letavocado** [comment]
For height alignment

```js
    //...

    if (barIndex > prevX) {
      const leftBarHeight = Math.round(prevLeft * halfHeight * barScale)
      const rightBarHeight = Math.round(prevRight * halfHeight * barScale)
      const barHeight = leftBarHeight + (rightBarHeight || 1)

      ctx.roundRect(
        prevX * (barWidth + barGap),
        options.alignEnd ? height - barHeight : halfHeight - leftBarHeight,
        barWidth,
        barHeight,
        barRadius
      )
    
    //...
```

p.s. upd

**Comment by katspaugh** [comment]
The proposed options makes sense. I suggest we call it `barAlign` though. E.g. `barAlign?: 'top' | bottom'`, wdyt?
Please feel free to make a PR for this.

**Reply by letavocado** [reply]
Thanks, @katspaugh! 
I'll made a PR with test. :+1: 

Cheers!

**Reply by letavocado** [reply]
![image](https://github.com/katspaugh/wavesurfer.js/assets/31704887/bc46d3ad-ed5e-49c8-858d-1db5d3498025)

I aligned by height, but I think that the highest bar should be equal to the canvas's height.

**Reply by katspaugh** [reply]
I guess `normalize: true` will take care of that once I restore it.

**Reply by letavocado** [reply]
PR: [2828](https://github.com/katspaugh/wavesurfer.js/pull/2828)

All local e2e tests failed, I will add tests later, as I figure it out. :thinking: 

![image](https://github.com/katspaugh/wavesurfer.js/assets/31704887/d9096bad-7624-4636-b039-81bc3138a754)

Cheers!

**Reply by katspaugh** [reply]
Nice, thank you!

**Comment by letavocado** [comment]
I guess we should add a warning if the bar's height greater than the canvas's height. What do you think?

**Reply by katspaugh** [reply]
It cannot be taller than the canvas height because it's always a magnitude value multiplied by the height. And the values are -1..1.

**Reply by letavocado** [reply]
If bar height is 5 it will.
![image](https://github.com/katspaugh/wavesurfer.js/assets/31704887/04561d30-20d4-4fcf-beef-03a6cf01eeee)

I agree with you. Besides, the extra check reduces the performance of the loop
